src/H5Iint.c

```diff
@@ -10195,263 +9746,54 @@ H5I__new_mt_type_info(const H5I_class_t *cls, unsigned reserved)
-
-                reallocable_entry_available = TRUE;
-            
-            } else { 
-
-                /* update stats */
-                atomic_fetch_add(&(H5I_mt_g.num_type_info_fl_num_reallocable_update_collisions), 1ULL);
-            }
-        } /* end while */
-#else
-        while ( ( ! reallocable_entry_available ) && 
-                ( 0 < (snum_fl_entries_reallocable = 
-                       atomic_load(&(H5I_mt_g.snum_type_info_fl_entries_reallocable))).val ) ) {
-
-            assert( 0 < snum_fl_entries_reallocable.val);
-
-            new_snum_fl_entries_reallocable.val = snum_fl_entries_reallocable.val - 1;
-            new_snum_fl_entries_reallocable.sn  = snum_fl_entries_reallocable.sn + 1;
-
-            if ( atomic_compare_exchange_strong(&(H5I_mt_g.snum_type_info_fl_entries_reallocable),
-                                                &snum_fl_entries_reallocable, new_snum_fl_entries_reallocable) ) {
-
-                reallocable_entry_available = TRUE;
-
-                fprintf(stderr, 
-                        "\nalloc: old/new snum_type_info_fl_entries_reallocable = {%lld, %lld} / {%lld, %lld}\n",
-                        (long long)snum_fl_entries_reallocable.val, 
-                        (long long)snum_fl_entries_reallocable.sn,
-                        (long long)new_snum_fl_entries_reallocable.val, 
-                        (long long)new_snum_fl_entries_reallocable.sn);
-            
-            } else { 
+                    assert( 0 < atomic_load(&(type_info_ptr->serial_num ) ) );
 
-                /* update stats */
-                atomic_fetch_add(&(H5I_mt_g.num_type_info_fl_num_reallocable_update_collisions), 1ULL);
-            }
-        } /* end while */
-#endif
-        if ( ! reallocable_entry_available ) {
-
-            /* No reallocable entries available, so set fl_search_done = TRUE */
-
-            fl_search_done = TRUE;
-
-            /* update stats */
-            atomic_fetch_add(&(H5I_mt_g.num_type_info_fl_alloc_req_denied_due_to_no_reallocable_entries), 1ULL);
-
-        }
-    } /* end if ( ! fl_search_done ) */
-
-
-    while ( ! fl_search_done ) {
-
-        sfirst = atomic_load(&(H5I_mt_g.type_info_fl_shead));
-        slast = atomic_load(&(H5I_mt_g.type_info_fl_stail));
-
-        assert(sfirst.ptr);
-        assert(slast.ptr);
-
-        snext = atomic_load(&(sfirst.ptr->fl_snext));
-
-        test_sfirst = atomic_load(&(H5I_mt_g.type_info_fl_shead));
-
-        if ( ( test_sfirst.ptr == sfirst.ptr ) && ( test_sfirst.sn == sfirst.sn ) ) {
-
-            if ( sfirst.ptr == slast.ptr ) {
+                    /* the above assert only exists in production builds.  If type_info_ptr->serial_num
+                     * is zero and we get this far, increment H5I_mt_g.num_id_info_fl_head_sn_is_zero.
+                     */
+                    if ( 0 == atomic_load(&(type_info_ptr->serial_num)) ) {
```

Comment:
Same as for ID, should probably throw an assertion failure for this right after read